src/include/duckdb/common/types/row/partitioned_tuple_data.hpp

```diff
@@ -0,0 +1,140 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/common/types/row/partitioned_tuple_data.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/common/perfect_map_set.hpp"
+#include "duckdb/common/types/row/tuple_data_allocator.hpp"
+#include "duckdb/common/types/row/tuple_data_collection.hpp"
+
+namespace duckdb {
+
+//! Local state for parallel partitioning
+struct PartitionedTupleDataAppendState {
+public:
+	PartitionedTupleDataAppendState() : partition_indices(LogicalType::UBIGINT) {
+	}
+
+public:
+	Vector partition_indices;
+	SelectionVector partition_sel;
+
+	static constexpr idx_t MAP_THRESHOLD = 32;
+	perfect_map_t<list_entry_t> partition_entries;
```

Comment:
Is a `perfect_map` not better implemented as a `vector`, or are the numbers here potentially very large?

Reply:
The usage so far almost always has small numbers but can in some cases have large numbers, but the size of the map (number of entries) will always be small. Therefore I chose not to use a vector. I want to get back to this at some point and unify computing the partition indices for both the `PartitionedColumnData` and `PartitionedTupleData`, and then I will try to find a more elegant solution.